File: architecture-skills/object-oriented-programming-paradigm/examples/single-responsibility/single-responsibility.swift
// Single Responsibility Principle in Swift.
//
// An Invoice that computes totals, formats itself, AND persists itself
// serves three actors. The refactor gives each actor its own type.

import Foundation

struct LineItem {
    let description: String
    let amount: Double
}

final class Invoice {
    let id: String
    let customer: String
    var items: [LineItem] = []

    init(id: String, customer: String) {
        self.id = id
        self.customer = customer
    }

    func addItem(description: String, amount: Double) {
        items.append(LineItem(description: description, amount: amount))
    }
}

struct InvoiceCalculator {
    let taxRate: Double

    func total(invoice: Invoice) -> Double {
        let subtotal = invoice.items.reduce(0) { $0 + $1.amount }
        return subtotal * (1 + taxRate)
    }
}

struct TextInvoiceFormatter {
    func format(invoice: Invoice, total: Double) -> String {
        var lines = ["Invoice \(invoice.id) for \(invoice.customer)"]
        for item in invoice.items {
            lines.append(String(format: "  %-12@ $%8.2f", item.description, item.amount))
        }
        lines.append(String(format: "  %-12@ $%8.2f", "TOTAL", total))
        return lines.joined(separator: "\n")
    }
}

final class InMemoryInvoiceRepository {
    private var rows: [String: Invoice] = [:]

    func save(_ invoice: Invoice) { rows[invoice.id] = invoice }
    func count() -> Int { rows.count }
}

let invoice = Invoice(id: "INV-042", customer: "ACME Corp")
invoice.addItem(description: "Consulting", amount: 1200)
invoice.addItem(description: "Hosting", amount: 90)

let calculator = InvoiceCalculator(taxRate: 0.20)
let formatter = TextInvoiceFormatter()
let repository = InMemoryInvoiceRepository()

let total = calculator.total(invoice: invoice)
print(formatter.format(invoice: invoice, total: total))

repository.save(invoice)
print("saved \(invoice.id) (\(repository.count()) record(s) stored)")

File: architecture-skills/object-oriented-programming-paradigm/examples/single-responsibility/single-responsibility.js
/** Single Responsibility Principle in JavaScript.
 *
 * An Invoice that computes totals, formats itself, AND persists itself
 * serves three actors. The refactor gives each actor its own collaborator.
 */

class LineItem {
  constructor(description, amount) {
    this.description = description;
    this.amount = amount;
  }
}

class Invoice {
  constructor(invoiceId, customer) {
    this.invoiceId = invoiceId;
    this.customer = customer;
    this.items = [];
  }

  addItem(description, amount) {
    this.items.push(new LineItem(description, amount));
  }
}

class InvoiceCalculator {
  constructor(taxRate) {
    this.taxRate = taxRate;
  }

  total(invoice) {
    const subtotal = invoice.items.reduce((sum, i) => sum + i.amount, 0);
    return subtotal * (1 + this.taxRate);
  }
}

class TextInvoiceFormatter {
  format(invoice, total) {
    const lines = [`Invoice ${invoice.invoiceId} for ${invoice.customer}`];
    for (const item of invoice.items) {
      lines.push(`  ${item.description.padEnd(12)} $${item.amount.toFixed(2).padStart(7)}`);
    }
    lines.push(`  ${'TOTAL'.padEnd(12)} $${total.toFixed(2).padStart(7)}`);
    return lines.join('\n');
  }
}

class InMemoryInvoiceRepository {
  constructor() {
    this.rows = new Map();
  }

  save(invoice) {
    this.rows.set(invoice.invoiceId, invoice);
  }

  count() {
    return this.rows.size;
  }
}

const invoice = new Invoice('INV-042', 'ACME Corp');
invoice.addItem('Consulting', 1200.00);
invoice.addItem('Hosting', 90.00);

const calculator = new InvoiceCalculator(0.20);
const formatter = new TextInvoiceFormatter();
const repository = new InMemoryInvoiceRepository();

const total = calculator.total(invoice);
console.log(formatter.format(invoice, total));

repository.save(invoice);
console.log(`saved ${invoice.invoiceId} (${repository.count()} record(s) stored)`);
